rust/opendp-ffi/src/any.rs

```diff
@@ -0,0 +1,593 @@
+    fn member(&self, val: &Self::Carrier) -> bool {
+        (self.member_glue)(self, val)
+    }
+}
+
+// TODO: If/when we remove the clone of the budget from make_adaptive_composition(), then remove Clone from AnyXXXDistance.
+#[derive(Clone, PartialEq)]
+pub struct AnyMeasureDistance {
+    distance: AnyBoxClonePartialEq,
+    partial_cmp_glue: Glue<fn(&Self, &Self) -> Option<Ordering>>,
+    sub_glue: Glue<fn(Self, &Self) -> Fallible<Self>>,
+}
+
+impl AnyMeasureDistance {
+    pub fn new<Q: 'static + Clone + MeasureDistance>(distance: Q) -> Self {
+        Self {
+            distance: AnyBoxClonePartialEq::new_clone_partial_eq(distance),
+            partial_cmp_glue: Glue::new(|self_: &Self, other: &Self| -> Option<Ordering> {
+                let self_ = self_.downcast_ref::<Q>().unwrap_assert("downcast of AnyMeasureDistance to constructed type will always work");
+                let other = other.downcast_ref::<Q>();
+                // FIXME: Do we want to have a FalliblePartialCmp for this?
+                other.map_or(None, |o| self_.partial_cmp(o))
+            }),
+            sub_glue: Glue::new(|self_: Self, rhs: &Self| -> Fallible<Self> {
+                let distance = self_.downcast::<Q>()?;
+                let rhs = rhs.downcast_ref::<Q>()?;
+                let res = distance.sub(rhs);
+                res.map(Self::new)
+            }),
+        }
+    }
+}
+
+impl Downcast for AnyMeasureDistance {
+    fn downcast<T: 'static>(self) -> Fallible<T> {
+        self.distance.downcast()
+    }
+    fn downcast_ref<T: 'static>(&self) -> Fallible<&T> {
+        self.distance.downcast_ref()
+    }
+}
+
+impl PartialOrd for AnyMeasureDistance {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        (self.partial_cmp_glue)(self, other)
+    }
+}
+
+impl FallibleSub<&Self> for AnyMeasureDistance {
+    type Output = Self;
+    fn sub(self, rhs: &Self) -> Fallible<Self::Output> {
+        // We have to clone sub_glue, because self is moved into the call.
+        self.sub_glue.clone()(self, rhs)
+    }
+}
+
+// TODO: If/when we remove the clone of the budget from make_adaptive_composition(), then remove Clone from AnyXXXDistance.
+#[derive(Clone, PartialEq)]
+pub struct AnyMetricDistance {
+    distance: AnyBoxClonePartialEq,
+    partial_cmp_glue: Glue<fn(&Self, &Self) -> Option<Ordering>>,
+}
+
+impl AnyMetricDistance {
+    pub fn new<Q: 'static + Clone + MetricDistance>(distance: Q) -> Self {
+        Self {
+            distance: AnyBoxClonePartialEq::new_clone_partial_eq(distance),
+            partial_cmp_glue: Glue::new(|self_: &Self, other: &Self| -> Option<Ordering> {
+                let self_ = self_.downcast_ref::<Q>().unwrap_assert("downcast of AnyMeasureDistance to constructed type will always work");
+                let other = other.downcast_ref::<Q>();
+                // FIXME: Do we want to have a FalliblePartialCmp for this?
+                other.map_or(None, |o| self_.partial_cmp(o))
```

Comment:
Not that this really matters, but this is possible:

```rust
let other = other.downcast_ref::<Q>().ok()?;
self_.partial_cmp(other)
```
Agreed that making fallible versions of existing traits to include downcasting errors is annoying. But PartialOrd is PartialOrd because the None variant is supposed to cover non-comparable values. On a different note, is it even possible to hit a PartialOrd downcast error after PartialEq has run on Domain/Metric/Measure?

Reply:
Nice, thanks!